wagon-provider-api/src/main/java/org/apache/maven/wagon/AbstractWagon.java

```diff
@@ -560,31 +564,78 @@ protected void transfer( Resource resource, InputStream input, OutputStream outp
     protected void transfer( Resource resource, InputStream input, OutputStream output, int requestType, long maxSize )
         throws IOException
     {
-        byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
+        byte[] buffer = bufferForTransferring( resource );
 
         TransferEvent transferEvent = new TransferEvent( this, resource, TransferEvent.TRANSFER_PROGRESS, requestType );
         transferEvent.setTimestamp( System.currentTimeMillis() );
 
         long remaining = maxSize;
         while ( remaining > 0 )
         {
-            // let's safely cast to int because the min value will be lower than the buffer size.
-            int n = input.read( buffer, 0, (int) Math.min( buffer.length, remaining ) );
+            // Read from the stream, block if necessary until either EOF or buffer is filled.
+            // Filling the buffer has priority since downstream processors will significantly degrade i/o
+            // performance if called to frequently (large data streams) as they perform expensive tasks such as
+            // console output or data integrity checks.
+            int nextByte = input.read();
 
-            if ( n == -1 )
+            if ( nextByte == -1 )
             {
                 break;
             }
 
-            fireTransferProgress( transferEvent, buffer, n );
+            buffer[0] = ( byte ) nextByte;
+
+            // let's safely cast to int because the min value will be lower than the buffer size.
+            int length = (int) min( buffer.length, remaining ),
+                read = 1;
+
+            for ( ; read < length ; ++read )
+            {
+                nextByte = input.read();
+                if ( nextByte == -1 )
+                {
+                    break;
+                }
+                buffer[read] = ( byte ) nextByte;
+            }
+
+            fireTransferProgress( transferEvent, buffer, read );
```

Comment:
I am somewhat confused why this performs better than described in `#read(byte[], int, int)`'s Javadoc::

> The read(b, off, len) method for class InputStream simply calls the method read() repeatedly ... The default implementation of this method blocks until the requested amount of input data len has been read, end of file is detected, or an exception is thrown. Subclasses are encouraged to provide a more efficient implementation of this method.

I understand that you are mimicking this behavior by filling the buffer manually byte-by-byte, but why is that better?!

Reply:
Makes sense. Have you considered passing to a `BufferInputStream` and see how it scales?

Reply:
I'd like to use standard components if possible.

Reply:
Hi @michael-o 

Yes, that was actually the first thing I did. However, BufferedInputStream does exactly what was done before in the loop, I.e. try and read *as many bytes as possible without blocking* into a buffer, thus resulting in millions of iterations for large files.

Reply:
Alright, to wrap up: BIS tries to read as many bytes as possible, but you are filling the buffer up to the top to avoid minichunking/fragmentation?

Reply:
Here is the snippet: https://github.com/JetBrains/jdk8u_jdk/blob/master/src/share/classes/sun/security/ssl/AppInputStream.java#L98-L113

So we hit to few `#avaliable()`, right?

Reply:
Yes, that's precisely it. I'm filling the buffer to avoid minichunking and indeed it's verly likely we always hit "to few" available() to make use of the entire buffer. Since the latter depends on the underlying network / hardware platform, I'm managing the buffer explicitly, prefering filling the buffer over non-blocking.

Reply:
Hi @michael-o ,
Is there anything else that stands in the way of this getting accepted / merged?